Merge the module package deployment into dev

lfshr opened the pull request:
Created logic to create module files and add modules to them as well as the deployment logic. -DeployModules now works and installs top to bottom!

Commit:
Working on modules WIP

Commit:
Changes to Modules - WIP

Commit:
Removed unworking code

Commit:
Merge branch 'dev' into module-dev

Commit:
Fixed module tools

Commit:
Merge branch 'dev' into module-dev

Commit:
Updated comment

Commit:
Cleanup and update to example

Commit:
Added dynamic parameter for module file to Add-AatPackageModule

Commit:
Updated dynamic parameter

Commit:
Removed unnecessary attribute from dynamic parameter

Commit:
Added help for modules stuff

stvnrs commented on docs/PackageManagement/Add-AatPackageModule.md:
```diff
@@ -0,0 +1,116 @@
+
+## SYNTAX
+
+### WithoutPackage (Default)
```
We need to avoid **Package** here, as its conflicts with our concept of package.  If we mean **BlobUri** we should probably say **BlobUri**

lfshr replied:
This will break all existing module files (I was aiming to have this just slot into place), but I'm happy to work this if you still wish.

lfshr replied:
Done.

stvnrs commented on docs/PackageManagement/Add-AatPackageModule.md:
````diff
@@ -0,0 +1,116 @@
+Accept wildcard characters: False
+```
+
+### -Package
````
We need to avoid **Package** here, as its conflicts with our concept of package.  If we mean **BlobUri** we should probably say **BlobUri**

lfshr replied:
As above.

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -573,6 +573,156 @@ function Add-AatPackageVariable {
+                [ordered]@{
+                    Name = 'AzureRm.Profile'
+                    Version = '2.7.0'
+                    Package = 'https://devopsgallerystorage.blob.core.windows.net/packages/azurerm.profile.2.7.0.nupkg'
```
We need to avoid **Package** here, as its conflicts with our concept of package.  If we mean **BlobUri** we should probably say **BlobUri**

lfshr replied:
As above

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -573,6 +573,156 @@ function Add-AatPackageVariable {
+                },
+                [ordered]@{
+                    Name = 'CustomPackage'
+                    Package = 'https://[storageaccountname].blob.core.windows.net/[ContainerName]/CustomPackage.zip'
```
we should enter a valid (but not real) sample path here.

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -573,6 +573,156 @@ function Add-AatPackageVariable {
+    param (
+        [Parameter(Mandatory = $true,
+                    Position = 2)]
+        [Parameter(ParameterSetName = 'WithoutPackage')]
```
We need to avoid **Package** here, as its conflicts with our concept of package.  If we mean **BlobUri** we should probably say **BlobUri**

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -573,6 +573,156 @@ function Add-AatPackageVariable {
+                    Position = 4,
+                    ParameterSetName = 'WithPackage')]
+        [string]
+        $Package
```
To be clearer this should be **BlobUri** or **StorageUri**

stvnrs replied:
or **ContentLink** like [New-​Azure​Rm​Automation​Runbook](https://docs.microsoft.com/en-us/powershell/module/azurerm.automation/new-azurermautomationrunbook?view=azurermps-1.7.0)

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -573,6 +573,156 @@ function Add-AatPackageVariable {
+        return $ParamDictionary
+    }
+
+    begin {
```
Without pipeline support don't use blocks.

lfshr replied:
This is needed as there is a dynamicparam block included. I have removed the unnecessary 'begin' block. The end block is still needed.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -767,27 +921,60 @@ function DeployModules {
+            try {
+                $ExistingModule = Get-AzureRmAutomationModule @Params
+            }
+            catch [Microsoft.Azure.Commands.Automation.Common.ResourceNotFoundException] {
```
Might be more 'natural' to use **ErrorAction** on the previous command rather than a try-catch.

lfshr replied:
The reason I'm using a try/catch here is so that I'm only catching the ResourceNotFoundException. There's no 'friendly' way to ask an Automation Account whether it has a module installed or not. This is a workaround. Would maybe be worth raising with the Azure team.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -767,27 +921,60 @@ function DeployModules {
+                $ExistingModule = $null
+            }
+
+            if ($null -eq $ExistingModule) {
```
Replace all with `if(-not $Thing)`

lfshr replied:
I try to avoid using the -not operator when testing against null as it's falsey. Can change this if we all agree.

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -767,27 +921,60 @@ function DeployModules {
+                    $AutomationModule.ProvisioningState -ne 'Succeeded' -and
+                    $AutomationModule.ProvisioningState -ne 'Failed') {
+                    Write-Error -Message "Failed to upload module: $($Module.Name) - PollCount reached max limit."
+                    continue
```
By default this should not continue on failure - it should throw. (this could be added later but we need to be fed through from **Publish**, so should be backlogged for now.

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -741,7 +891,11 @@ function DeployModules {
+
+        [Parameter()]
+        [ValidateNotNullOrEmpty()]
+        [int]$MaxModulePoll = 100
```
This should be clear that this is a max number of attempts e.g. **MaxPollAttempts**.  The Poll interval should also be an optional parameter e.g. **PollIntervalSeconds**

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -767,27 +921,60 @@ function DeployModules {
-                Write-Output "Adding new module - done."
+                if ($AutomationModule.ProvisioningState -eq 'Failed') {
+                    Write-Error -Message "Failed to upload module: $($Module.Name) - ProvisioningState Failed"
+                    continue
```
By default this should not continue on failure - it should throw. (this could be added later but we need to be fed through from **Publish**, so should be backlogged for now.

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -898,3 +1085,186 @@ function Publish-AatAutomationPackage {
+        $FinalList
+    )
+
+    begin {
```
Do not use blocks except for pipleline support

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -898,3 +1085,186 @@ function Publish-AatAutomationPackage {
     }
 }
+
+function GetModuleBlob {
```
Add comment with reference to the project/module that inspired this.

lfshr replied:
Sorry, this was in the original. I must have forgot to pull it over.

lfshr replied:
Done.

stvnrs reviewed (changes requested):
Quite a bit to address before we can merge this.

Commit:
Changed example on New-AatModulesFile

Commit:
Removed script blocks from cmdlet

lfshr commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -573,6 +573,156 @@ function Add-AatPackageVariable {
+        return $ParamDictionary
+    }
+
+    begin {
```
Done.

Commit:
Changed module poll variables

Commit:
Changed non-terminating errors to terminating errors.

Commit:
Removed unnecessary begin block

Commit:
Renamed package to ContentUri in module files

Commit:
Changed ContentUri to ContentLink

Commit:
Added credits to functions

Commit:
Changed '$null -eq' to '-not'

Commit:
Removed aliases

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -573,6 +573,156 @@ function Add-AatPackageVariable {
+        $Content = [System.Collections.Generic.List[hashtable]]::new()
+
+        if ($IncludeSamples.IsPresent) {
+            Write-Output "`n`tAzureRm.Profile has been added to the example using the `"Add-AatPackageModule -Name 'AzureRM.Profile' -Version '2.7.0'`" cmdlet." | Out-Default
```
Should be a Write-Verbose.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -573,6 +573,156 @@ function Add-AatPackageVariable {
+        $ValidateNotNullOrEmptyAttribute = New-Object System.Management.Automation.ValidateNotNullOrEmptyAttribute
+
+        $ValidateSetAttribute = New-Object -TypeName System.Management.Automation.ValidateSetAttribute -ArgumentList (
+            Get-ChildItem -Path (Get-AatPackageFolderPath -Modules) -Filter '*.json' | %{$_.Name -replace '.json', ''}
```
Probably shouldn't assume the user hasn't supplied .json on the file name.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -573,6 +573,156 @@ function Add-AatPackageVariable {
+        [Parameter(ParameterSetName = 'WithoutPackage')]
+        [Parameter(ParameterSetName = 'WithPackage')]
+        [string]
+        $Name,
```
We should probably validated that is *.json - which should also be added to **New-AatAssetsFile**

Optionally we could append it if its not specified.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -573,6 +573,156 @@ function Add-AatPackageVariable {
+        $Package
+    )
+
+    dynamicparam{
```
Should probably raise an issue so that Add-AatPackageVariable has equivalent functionality.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -617,27 +617,28 @@ function New-AatModulesFile {
         [Parameter(Mandatory = $true,
                     Position = 4,
-                    ParameterSetName = 'WithPackage')]
+                    ParameterSetName = 'PSGalleryModule
```
This `ParameterSetName` should be `CustomModule`?

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -617,27 +617,28 @@ function New-AatModulesFile {
+        ')]
         [string]
-        $Package
+        $ContentLink
```
This should have basic validation that this is a uri.

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -683,14 +682,14 @@ function Add-AatPackageModule {
             ConvertFrom-Json
 
-        if ($PSCmdlet.ParameterSetName -eq 'WithoutPackage') {
+        if ($PSCmdlet.ParameterSetName -eq 'CustomModule') {
```
This looks like it should be if `ParameterSetName -eq PSGalleryModule'

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -889,13 +888,18 @@ function DeployAssets {
+
+        # The time in seconds to wait between poll attempts
+        [Parameter()]
+        [int]$PollInterval = 10
```
`PollIntervalSeconds`

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -952,24 +956,24 @@ function DeployModules {
                     $AutomationModule.ProvisioningState -ne 'Failed') {
-                    Write-Error -Message "Failed to upload module: $($Module.Name) - PollCount reached max limit."
+                    throw "Failed to upload module: $($Module.Name) - PollCount reached max limit."
                     continue
```
`Continue`  is redundant after `throw`

lfshr replied:
Done.

stvnrs commented on src/PackageManagement/AzureAutomationTools.PackageManagement.psm1:
```diff
@@ -952,24 +956,24 @@ function DeployModules {
                 if ($AutomationModule.ProvisioningState -eq 'Failed') {
-                    Write-Error -Message "Failed to upload module: $($Module.Name) - ProvisioningState Failed"
+                    throw "Failed to upload module: $($Module.Name) - ProvisioningState Failed"
                     continue
```
`continue` is redundant after `throw`

lfshr replied:
Done.

Commit:
Fixed ParameterSet bug

Commit:
Added validation on ContentLink
Added validation to make sure ContentLink is infact a valid Uri. Content link must start with 'https://' or 'http://', have a body, then a period with a body, then a forward slash and a body at the end.

Commit:
Renamed PollInterval to PollIntervalSeconds

Commit:
Removed redundant continue

lfshr commented:
@stvnrs sorry, forgot to say there are updates to this module awaiting review.
